src/main/java/io/victoriametrics/client/metrics/Summary.java

```diff
@@ -0,0 +1,170 @@
+    }
+
+    private void validateQuantiles(double[] quantiles) {
+        for (double quantile : quantiles) {
+            if (quantile < 0 || quantile > 1 ) throw new IllegalArgumentException("Quantile must be in range 0 and 1");
+        }
+    }
+
+    public void update(double value) {
+        sum.add(value);
+        count.increment();
+        timeWindowQuantile.insert(value);
+    }
+
+    public double getSum() {
+        return sum.doubleValue();
+    }
+
+    public long getCount() {
+        return count.longValue();
+    }
+
+    /**
+     * Get the estimated value at the specified quantile.
+     * @param quantile Requested quantile
+     */
+    public double getQuantil(double quantile) {
+        return timeWindowQuantile.get(quantile);
+    }
+
+    /**
+     * Get estimated values by configured quantiles.
+     */
+    public SortedMap<Double, Double> getQuantilValues() {
+        SortedMap<Double, Double> result = new TreeMap<>();
+        for (double quantile : quantiles) {
+            result.put(quantile, timeWindowQuantile.get(quantile));
+        }
+
+        return result;
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public void accept(MetricVisitor visitor) {
+        visitor.visit(this);
+    }
+
+    private static class TimeWindowQuantile {
+        private final TimeWindow[] timeWindow;
+
+        private final long rotationDurationMillis;
+
+        private long lastRotationTimestamp;
+
+        private int currentWindow = 0;
+
+        private TimeWindowQuantile(long windowDurationSeconds, int timeWindows) {
+            this.timeWindow = new TimeWindow[timeWindows];
+            for (int i = 0; i < timeWindows; i++) {
+                timeWindow[i] = new TimeWindow();
+            }
+
+            this.rotationDurationMillis = TimeUnit.SECONDS.toMillis(windowDurationSeconds) / timeWindows;
+        }
+
+        public synchronized void insert(double value) {
+            rotate();
+            for (TimeWindow timeFrame : timeWindow) {
+                timeFrame.insert(value);
+            }
+        }
+
+        public synchronized double get(double phi) {
+            TimeWindow window = rotate();
+            return window.get(phi);
+        }
+
+        public TimeWindow rotate() {
+            long elapsedFromLastRotation = System.currentTimeMillis() - lastRotationTimestamp;
+
+            while(elapsedFromLastRotation > rotationDurationMillis) {
+                timeWindow[currentWindow] = new TimeWindow();
+                if (++currentWindow >= timeWindow.length) {
+                    currentWindow = 0;
+                }
+
+                elapsedFromLastRotation -= rotationDurationMillis;
+                lastRotationTimestamp += rotationDurationMillis;
+            }
+
+            return timeWindow[currentWindow];
+        }
+    }
+
+    private static class TimeWindow {
+
+        private final List<Double> samples = new LinkedList<>();
```

Comment:
There are very few reasons to use LinkedList (even its author joked that he haven't used it themselves). Why not ArrayList?